psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1412 @@
+    localizer->set_maxiter(options_.get_int("LOCAL_MAXITER"));
+    localizer->localize();
+
+    C_lmo = localizer->L();
+    F_lmo = linalg::triplet(C_lmo, reference_wavefunction_->Fa(), C_lmo, true, false, false);
+
+    // Form projected atomic orbitals by removing occupied space from the basis
+    C_pao = std::make_shared<Matrix>("Projected Atomic Orbitals", nbf, nbf);
+    C_pao->identity();
+    C_pao->subtract(linalg::triplet(C_occ, C_occ, reference_wavefunction_->S(), false, true, false));
+    S_pao = linalg::triplet(C_pao, reference_wavefunction_->S(), C_pao, true, false, false);
+
+    // normalize PAOs
+    for (size_t i = 0; i < C_pao->colspi(0); ++i) {
+        C_pao->scale_column(0, i, pow(S_pao->get(i, i), -0.5));
+    }
+    S_pao = linalg::triplet(C_pao, reference_wavefunction_->S(), C_pao, true, false, false);
+    F_pao = linalg::triplet(C_pao, reference_wavefunction_->Fa(), C_pao, true, false, false);
+
+    // map from atomic center to orbital/aux basis function/shell index
+
+    atom_to_bf_.resize(natom);
+    atom_to_ribf_.resize(natom);
+    atom_to_shell_.resize(natom);
+    atom_to_rishell_.resize(natom);
+
+    for (size_t i = 0; i < nbf; ++i) {
+        atom_to_bf_[basisset_->function_to_center(i)].push_back(i);
+    }
+
+    for (size_t i = 0; i < naux; ++i) {
+        atom_to_ribf_[ribasis_->function_to_center(i)].push_back(i);
+    }
+
+    for (size_t s = 0; s < nshell; s++) {
+        atom_to_shell_[basisset_->shell_to_center(s)].push_back(s);
+    }
+
+    for (size_t s = 0; s < nshellri; s++) {
+        atom_to_rishell_[ribasis_->shell_to_center(s)].push_back(s);
+    }
+}
+
+double DLPNOMP2::compute_energy() {
+    timer_on("DLPNO-MP2");
+
+    print_header();
+
+    setup();
+
+    timer_on("Overlap Ints");
+    overlap_ints();
+    timer_off("Overlap Ints");
+
+    timer_on("Dipole Ints");
+    dipole_ints();
+    timer_off("Dipole Ints");
+
+    timer_on("Sparsity");
+    sparsity_prep();
+    timer_off("Sparsity");
+
+    timer_on("DF Ints");
+    df_ints();
+    timer_off("DF Ints");
+
+    timer_on("PNO Transform");
+    pno_transform();
+    timer_off("PNO Transform");
+
+    timer_on("PNO Overlaps");
+    pno_overlaps();
+    timer_off("PNO Overlaps");
+
+    timer_on("LMP2");
+    lmp2_iterations();
+    timer_off("LMP2");
+
+    print_results();
+
+    timer_off("DLPNO-MP2");
+
+    double e_scf = reference_wavefunction_->energy();
+    double e_mp2_corr = e_lmp2_ + de_dipole_ + de_pno_total_;
+    double e_mp2_total = e_scf + e_mp2_corr;
+
+    set_scalar_variable("MP2 CORRELATION ENERGY", e_mp2_corr);
+    set_scalar_variable("CURRENT CORRELATION ENERGY", e_mp2_corr);
+    set_scalar_variable("MP2 TOTAL ENERGY", e_mp2_total);
+    set_scalar_variable("CURRENT ENERGY", e_mp2_total);
+
+    set_scalar_variable("MP2 SINGLES ENERGY", 0.0);
+    set_scalar_variable("MP2 DOUBLES ENERGY", e_mp2_corr);
+    set_scalar_variable("MP2 SAME-SPIN CORRELATION ENERGY", 0.0);      // TODO
```

Comment:
Lori, should these variables even be set?

Reply:
follow this pattern, I think: https://github.com/psi4/psi4/blob/master/psi4/src/psi4/dfmp2/mp2.cc#L736-L750 . If dlpno doesn't decompose readily into SS & OS, then don't set those variables.

the block you have here will get them set on the wfn. then in proc.py add this block https://github.com/psi4/psi4/blob/master/psi4/driver/procrouting/proc.py#L3932-L3934 so that global P::e has the same info.